Tests/CADCacheAdvanceTests/CADCacheAdvanceTests.m

```diff
@@ -0,0 +1,616 @@
+{
+    CADCacheAdvance *const cache = [self createCacheThatOverwitesOldMessages:YES];
+    XCTAssertTrue(cache.isEmpty);
+}
+
+- (void)test_isEmpty_returnsFalseForANonEmptyCache;
+{
+    CADCacheAdvance *const cache = [self createCacheThatOverwitesOldMessages:YES];
+    NSError *error = nil;
+    [cache appendMessage:[@"Test" dataUsingEncoding:NSUTF8StringEncoding]
+                error:&error];
+    XCTAssertNil(error);
+    XCTAssertFalse(cache.isEmpty);
+}
+
+- (void)test_messagesAndReturnError_returnsAppendedMessage;
+{
+    CADCacheAdvance *const cache = [self createCacheThatOverwitesOldMessages:YES];
+    NSData *const message = [@"Test" dataUsingEncoding:NSUTF8StringEncoding];
+    NSError *error = nil;
+    [cache appendMessage:message
+                   error:&error];
+    XCTAssertNil(error);
+    XCTAssertEqualObjects([cache messagesAndReturnError:nil], @[message]);
+}
+
+// MARK: Performance Tests
+
+- (void)test_performance_append_fillableCache;
+{
+    CADCacheAdvance *const cache = [self createCacheThatOverwitesOldMessages:NO];
+    // Fill the cache before the test starts.
+    for (NSString *const message in [self lorumIpsumMessages]) {
+        [cache appendMessage:[message dataUsingEncoding:NSUTF8StringEncoding]
+                       error:nil];
+    }
+
+    [self measureBlock:^{
+        for (NSString *const message in [self lorumIpsumMessages]) {
+            [cache appendMessage:[message dataUsingEncoding:NSUTF8StringEncoding]
+                           error:nil];
+        }
+    }];
+}
+
+- (void)test_performance_messages_fillableCache;
+{
+    CADCacheAdvance *const cache = [self createCacheThatOverwitesOldMessages:NO];
+    // Fill the cache before the test starts.
+    for (NSString *const message in [self lorumIpsumMessages]) {
+        [cache appendMessage:[message dataUsingEncoding:NSUTF8StringEncoding]
+                       error:nil];
+    }
+
+    [self measureBlock:^{
+        (void)[cache messagesAndReturnError:nil];
+    }];
+}
+
+// MARK: Private
+
+- (CADCacheAdvance *)createCacheThatOverwitesOldMessages:(BOOL)overwritesOldMessages;
+{
+    [NSFileManager.defaultManager createFileAtPath:[self testFileLocation].path
+                                          contents:nil
+                                        attributes:nil];
+    NSError *error = nil;
+    CADCacheAdvance *const cache = [[CADCacheAdvance alloc]
+                                    initWithFileURL:self.testFileLocation
+                                    maximumBytes:1000000
+                                    shouldOverwriteOldMessages:overwritesOldMessages
+                                    error:&error];
+    XCTAssertNil(error, "Failed to create cache due to %@", error);
+    return cache;
+}
+
+- (NSURL *)testFileLocation;
+{
+    return [NSFileManager.defaultManager.temporaryDirectory URLByAppendingPathComponent:@"CADCacheAdvanceTests"];
+}
+
+- (NSArray<NSString *> *)lorumIpsumMessages;
```

Comment:
What do you think about putting these in a file that we can read from disk in both sets of tests?
It would be nice if we had a stronger guarantee that they were in sync so that we are sure we're comparing apples to apples when we're comparing performance test results.
Another option would be to put them in their own target that we share between tests.

Reply:
I can give that a shot

Reply:
Gave it a shot in https://github.com/dfed/CacheAdvance/pull/42/commits/dc3137e7472a4c654302ef6490648329dc5f0d63. I may need to fix up some CI issues, but should be doable. Solid idea 🙂

Reply:
That worked out great, and much DRY-er!